neps/nep-0413.md

```diff
@@ -0,0 +1,162 @@
+
+### Why The Message Needs to Include a Service Identifier and Nonce?
+To stop a malicious app from requesting the user to sign a message for them, only to relay it to a third-party. Including the recipient and making sure the user knows about it should mitigate this kind of attacks.
+
+Meanwhile, including a nonce helps to mitigate replay attacks, in which an attacker can delay or re-send a signed message.
+
+### Why using a FullAccess Key? Why Not Simply Creating an [FunctionCall Key](https://docs.near.org/concepts/basics/accounts/access-keys) for Signing?
+The most common flow for [NEAR user authentication into a Web3 frontend](https://docs.near.org/develop/integrate/frontend#user-sign-in--sign-out) involves the creation of a [FunctionCall Key](](https://docs.near.org/concepts/basics/accounts/access-keys)).
+
+One might feel tempted to reproduce such process here, for example, by creating a key that can only be used to call a non-existing method in the user's account. This is a bad idea because:
+1. The user would need to expend gas in creating a new key.
+2. Any third-party can ask the user to create a `FunctionCall Key`, thus opening an attack vector.
+
+Using a FullAccess key allows us to be sure that the challenge was signed by the user (since nobody should have access to their `FullAccess Key`), while keeping the constraints of not expending gas in the process (because no new key needs to be created).
+
+### How to Return the Signed Message in a Safe Way
+Sending the signed message in a query string to an arbitrary URL (even within the correct domain) is not secure as the data can be leaked (e.g. through headers, etc). Using URL fragments instead will improve security, since [URL fragments are not included in the `Referer`](https://greenbytes.de/tech/webdav/rfc2616.html#header.referer).
+
+### NEAR Signatures
+NEAR transaction signatures are not plain Ed25519 signatures but Ed25519 signatures of a SHA-256 hash (see [near/nearcore#2835](https://github.com/near/nearcore/issues/2835)). Any protocol that signs anything with NEAR account keys should use the same signature format.
+
+## Specification
+Wallets must implement a `signMessage` method, which takes a `message` destined to a specific `recipient` and transform it into a verifiable signature.
+
+### Input Interface
+`signMessage` must implement the following input interface:
+
+```jsx
+interface SignMessageParams {
+  message: string ; // The message that wants to be transmitted.
+  recipient: string; // The recipient to whom the message is destined (e.g. "alice.near" or "myapp.com").
+  nonce: [u8; 32] ; // A nonce that uniquely identifies this instance of the message, denoted as a 32 bytes array (a fixed `Buffer` in JS/TS).
+  callbackUrl?: string; // Optional, applicable to browser wallets (e.g. MyNearWallet). The URL to call after the signing process. Defaults to `window.location.href`.
+}
```

Comment:
@abacabadabacaba if you go through the +100 messages above you will see that the NEP is a result of what the community wanted: a way to sign a string message for a specific receiver (e.g. "anna.near", "myweb.com", or "aunt Mary"). Since some people wanted to use this as a way of login-while-passing-data (see the discussion), a `nonce` was added, so the same message for the same receiver could not be used twice. 

While I understand the convenience of having two methods separated, this is not what the community was asking for. If there is no major drawbacks, I would suggest to keep it like this.

If this is a major drawback, please make a PR with the suggested changes so I can review it and maybe we can create another NEP together.

Reply:
Until we get more details on why `recipient`, `nonce` and `callbackUrl` are not needed for authenticating a human-readable message for another person the only action item I see here is to elaborate on the role of the `recipient` field perhaps providing some examples as well.

I agree that it's preferable to have a more flexible standard that can be adapted to both use cases if there are no major security vulnerabilities that implementors of this standard are likely to face.